src/synopsys-action/factory/github-client-service-factory.ts

```diff
@@ -0,0 +1,60 @@
+import {GithubClientServiceInterface} from '../service/github-client-service-interface'
+import {GithubClientServiceCloud} from '../service/impl/cloud/github-client-service-cloud'
+import {info} from '@actions/core'
+import * as constants from '../../application-constants'
+import {GithubClientServiceV1} from '../service/impl/enterprise/v1/github-client-service-v1'
+import {GithubClientServiceV2} from '../service/impl/enterprise/v2/github-client-service-v2'
+import {HttpClient} from 'typed-rest-client/HttpClient'
+import * as inputs from '../inputs'
+
+export const GitHubClientServiceFactory = {
+  SUPPORTED_VERSIONS_V1: ['3.11', '3.12'],
+  SUPPORTED_VERSIONS_V2: ['3.13', '3.14'],
+  DEFAULT_VERSION: '3.12',
+
+  async fetchVersion(githubServerUrl: string): Promise<string> {
+    const githubEnterpriseMetaUrl = '/meta'
+    const githubToken = inputs.GITHUB_TOKEN
+    const endpoint = githubServerUrl.concat(githubEnterpriseMetaUrl)
+
+    try {
+      const httpClient = new HttpClient('GitHubClientServiceFactory')
+      const httpResponse = await httpClient.get(endpoint, {
+        Authorization: `Bearer ${githubToken}`,
+        Accept: 'application/vnd.github+json'
+      })
+
+      if (httpResponse.message.statusCode === constants.HTTP_STATUS_OK) {
+        const metaDataResponse = JSON.parse(await httpResponse.readBody())
+        return metaDataResponse.installed_version
+      } else {
+        throw new Error(`No version info found for endpoint : ${endpoint}`)
+      }
+    } catch (error) {
+      throw new Error(`Fetching version info for enterprise server failed : ${error}`)
+    }
+  },
+
+  async getGitHubClientServiceInstance(): Promise<GithubClientServiceInterface> {
+    info('Fetching GitHub client service instance...')
+    const githubServerUrl = process.env[constants.GITHUB_ENVIRONMENT_VARIABLES.GITHUB_SERVER_URL] || ''
```

Comment:
minor: To keep it consistent with adapters, use GITHUB_API_URL.

Reply:
@lsynopsys Are you suggesting to rename the env variable name for  `GITHUB_SERVER_URL` to `GITHUB_API_URL`?
We are using `GITHUB_SERVER_URL` reference in many places in the code.

Current env variables for GitHub that we use are - 

```
export const GITHUB_ENVIRONMENT_VARIABLES = {
  GITHUB_TOKEN: 'GITHUB_TOKEN',
  GITHUB_REPOSITORY: 'GITHUB_REPOSITORY',
  GITHUB_HEAD_REF: 'GITHUB_HEAD_REF',
  GITHUB_REF: 'GITHUB_REF',
  GITHUB_REF_NAME: 'GITHUB_REF_NAME',
  GITHUB_REPOSITORY_OWNER: 'GITHUB_REPOSITORY_OWNER',
  GITHUB_BASE_REF: 'GITHUB_BASE_REF',
  GITHUB_EVENT_NAME: 'GITHUB_EVENT_NAME',
  GITHUB_SERVER_URL: 'GITHUB_SERVER_URL',
  GITHUB_SHA: 'GITHUB_SHA'
}
```

Reply:
create a new variable as `GITHUB_API_URL` which will refer to h`ttps://api.github.com`.
ref: https://docs.github.com/en/actions/learn-github-actions/contexts

Reply:
Updated accordingly. Please check.